Checkout: add stripe elements form

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This adds a new credit card checkout form, currently controlled by AB test settings on the server (not visible to most users yet), which uses Stripe Elements and the new Payment Methods API. The forms intentionally look very similar.

![regular-form](https://user-images.githubusercontent.com/2036909/61395600-d644c200-a893-11e9-80f2-ea351847a7d9.png)

![stripe-form](https://user-images.githubusercontent.com/2036909/61395609-d8a71c00-a893-11e9-9dff-40418a067b33.png)

To do:

- [x] Make sure the transactions endpoint accepts the payment method.
- [x] Match the design of the CC field to the other fields (eg: thin border).
- [x] Disable the submit button when Stripe or the CC field are loading.
- [x] Fetch the Stripe API key from an endpoint.
- [x] Determine where to store the Stripe JS url fetched by the page.
- [x] Add handling for when the transactions endpoint returns various responses.
- [x] Add country form field.
- [x] Use ~country form field~ geo-ip to toggle Stripe API key.
- [x] Add subscription length toggle to form (1 year, 2 years) and make sure it changes the page (price, etc.).
- [x] Show validation errors on the stripe fields.

#### Testing
1. To test the Stripe.js based CC fields, enable store sandbox mode.
2. To test the existing paygate based CC fields with this PR, enable store sandbox mode, but disallow the `WPCOM_Billing_Stripe_Payment_Method` class in the shopping-cart allowed methods .

#### Test cases
##### Stripe / New Cards
- [x] 1. Auth never required, charge success - `4242424242424242`
**Expected result:** transaction succeeds => redirected to the success page 

- [x] 2. Auth not required, charge fails  - `4000000000000002`
**Expected result:** transaction fails => an error message is displayed.

- [x] 3. Auth required (always), auth succeeds, charge succeeds  - `4000002760003184`
**Expected result:** auth succeeds => transaction succeeds => redirected to "processing" page => redirected to success page

- [x] 4. Auth required (first payment only), auth succeeds, charge succeeds  - `4000002500003155`
**Expected result:** auth succeeds => transaction succeeds => redirected to "processing" page => redirected to success page

- [x] 5. Auth required (always), choose to fail auth  - `4000002760003184`
**Expected result:** auth fails => an error message is displayed.

- [x] 6. Auth required, auth succeeds, charge fails  - `4000008260003178`
**Expected result:** auth succeeds => transaction fails => an error message is displayed.

##### Stripe / Stored Cards
- [x] 7. Auth never required, charge success - reuse card from test **1**
**Expected result:** transaction succeeds => redirected to the success page 

- [x] 8. Auth required (off-session only for first payment), charge succeeds - reuse card from test **4**
**Expected result:** auth succeeds => transaction succeeds => redirected to "processing" page => 
redirected to success page

- [x] 9. Auth required, auth succeeds, charge succeeds - reuse card from test **3**
**Expected result:** auth succeeds => transaction succeeds => redirected to "processing" page => 
redirected to success page

- [x] 10. Auth required, choose to fail auth   - reuse card from test **3**
**Expected result:** auth fails => an error message is displayed.

##### Paygate / New Cards
- [x] 11. Auth never required, charge success - `4242424242424242`
**Expected result:** transaction succeeds => redirected to the success page 

- [x] 12. Auth not required, charge fails  - `4000000000000002`
**Expected result:** transaction fails => an error message is displayed.

- [x] 13. 3DS Auth required (always), auth succeeds, charge succeeds  - `4000000000003063`
**Expected result:** Charge fails (transparent) => 3DS Redirect => 3DS Approved => redirected to "processing" page => Transaction succeeds => redirected to success page

- [x] 14. 3DS Auth required (always), choose to fail auth, charge fails  - `4000000000003063`
**Expected result:** Charge fails (transparent) => 3DS Redirect => 3DS Declined => redirected to chekcout page

- [x] ~~15. 3DS Auth required (always), auth succeeds, charge fails  - `4000008400001629`~~ No card available for this scenario
**Expected result:** Charge fails (transparent) => 3DS Redirect => 3DS Approved => redirected to "processing" page => Transaction fails => redirected to checkout page with error message

##### Paygate / Saved Cards
- [x] 16. Auth never required, charge success - reuse card from test **11**
**Expected result:** transaction succeeds => redirected to the success page 

- [x] 17. Auth required, auth succeeds, charge succeeds - reuse card from test **13**
**Expected result:** Charge fails (transparent) => 3DS Redirect => 3DS Approved => redirected to "processing" page => Transaction succeeds => redirected to success page

- [x] 18. 3DS Auth required (always), choose to fail auth, charge fails - reuse card from test **13**
**Expected result:** Charge fails (transparent) => 3DS Redirect => 3DS Declined => redirected to chekcout page

a8ci18n commented on client/my-sites/checkout/checkout/stripe-elements-payment-box.jsx:
```diff
@@ -0,0 +1,216 @@
+			span: <span className="credit-card-form-fields__explainer" />,
+		},
+	} );
+	const cardDetailsLabel = translate( 'Card Details' );
```
Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Edit Card Details', { context: 'Header text'} )` *ES Score: 7*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/stripe-elements-checkout)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/checkout/checkout/stripe-elements-payment-box.jsx:
```diff
@@ -0,0 +1,216 @@
+		},
+	} );
+	const cardDetailsLabel = translate( 'Card Details' );
+	const payButtonLabel = translate( 'Pay %(price)s', {
```
Hi! I've found a possible matching string that has already been translated **51** times:
`translate( 'Pay %(price)s', { context: 'Pay button on /checkout'} )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

yoavf commented on client/my-sites/checkout/checkout/stripe-elements-payment-box.jsx:
```diff
@@ -0,0 +1,260 @@
+					);
+					window.location.href = result.redirect_url;
+				}
+				// TODO: what if there is no redirect_url?
```
We won't be using the redirect_Url  in this payment box since we won't be redirecting to an external URL. Stripe.js will use the `message` property we'll pass to `handleStripeAction` to take care of the authentication.

However, if there's no `message` property (and we get a success:true, and a receipt id), then we have a successful CC payment that doesn't require authentication, and we just need to redirect to the success page.

yoavf replied:
At this point, I'm not sure if it's maybe worth to reuse the transactionSteps/submitTransaction logic from the regular cc form

yoavf commented on client/my-sites/checkout/checkout/stripe-elements-payment-box.jsx:
```diff
@@ -0,0 +1,268 @@
+import { isWpComBusinessPlan, isWpComEcommercePlan } from 'lib/plans';
+import PaymentChatButton from './payment-chat-button';
+import SubscriptionText from './subscription-text';
+import PaymentCountrySelect from 'components/payment-country-select';
```
I'm wondering if we could drop this:
> The card Element automatically determines your customer’s billing address country based on their card number. 

Maybe we can fill the country based on the card?

sirbrillig replied:
I think we should definitely do that. I'll just have to figure out how 😁

sirbrillig commented on client/my-sites/checkout/checkout/redirect-payment-box.jsx:
```diff
@@ -124,16 +124,12 @@ export class RedirectPaymentBox extends PureComponent {
 
-	redirectToPayment = event => {
-		const origin = getLocationOrigin( location );
+	redirectToPayment = async event => {
```
Yeah. This was mostly for the purpose of understanding the existing flow so I could copy it. It has been moved to another branch for now.

sirbrillig committed:
SecurePaymentForm: make event optional in submitTransaction

The only thing that `event` is used for here is to call
`preventDefault()`, and if we are using async synthetic React events,
then `preventDefault()` may already have been called, making it
challenging to pass the un-cancelled event to this function. With this
change, the argument becomes optional.

sirbrillig commented:
Refactored to use `TransactionFlow` for submission. This still needs some work on validation. For example, if you do not enter a credit card number, it appears that the form submits but nothing happens because we are not showing the validation error.

sirbrillig commented:
Oh! Thank you for fixing the styles @yoavf! I spent way too long trying to get those fields to look right.

sirbrillig commented:
Refactored to show validation errors and to fetch the api key from the server (requires D30074-code).

sirbrillig commented:
I re-added the call to `stripe.handleCardPayment()` ([docs](https://stripe.com/docs/stripe-js/reference#stripe-handle-card-payment)) which @yoavf had added before I rebased this branch to use `TransactionFlow`. However, I removed the redirect to the `/pending/:orderId` page because I'm not sure I understand what that was supposed to do. That redirect happened after `handleCardPayment()` was complete; I think it was either supposed to happen during the call in case a 3d Secure popup was shown and was in the wrong place, or something else that I don't grok.

sirbrillig commented:
If this works correctly now, we should show the Stripe Elements form masquerading as the regular credit-card form whenever `SecurePaymentForm` is displayed and the allowed payment methods returned by the server include `WPCOM_Billing_Stripe_Payment_Method` as an option.

yoavf commented:
> However, I removed the redirect to the `/pending/:orderId` page because I'm not sure I understand what that was supposed to do. That redirect happened after `handleCardPayment()` was complete; I think it was either supposed to happen during the call in case a 3d Secure popup was shown and was in the wrong place, or something else that I don't grok.

So there are a few response options when paying with a `paymentIntent`:

1. Success
The card didn't require authentication, payment succeeded immediately and provisioned by the backend => complete payment, redirect to success page (maybe through `Checkout.handleCheckoutCompleteRedirect` but I'm not sure)

2. Failure
The payment failed immediately => transaction fail, display error messgae

3. Authentication required (based on the existence of result.message in `_submitWithPayment` call):
We call `handleCardPayment()` which will (in most cases) show a 3D Secure authentication modal. If `handleCardPayment()` contains result.error = then we show an error message. Otherwise, we assume the payment is correct and we need to redirect to the `/pending/:orderId` page because provisioning is done on the backend using a webhook. That page will handle the final redirect.

yoavf commented:
I've just pushed changes that should handle all of the scenarios from my previous message.

yoavf commented on test/e2e/lib/components/secure-payment-component.js:
```diff
@@ -42,6 +42,24 @@ export default class SecurePaymentComponent extends AsyncBaseContainer {
+			'Could not locate the ElementInput iFrame.'
+		);
+
+		this.driver.getPageSource().then( function( source ) {
```
debugging code, need to be removed

sirbrillig commented:
I made some adjustments including sending the `payment_partner` from the stripe configuration endpoint response to the transactions endpoint. This requires D30485-code and D30490-code. Currently the stripe configuration endpoint (with those changes) calls that field `processor_id` rather than `payment_partner` since it's already said that the payment partner is Stripe, but maybe it should have the same name in both places. What do you think?

yoavf commented:
That sounds good - and good call on passing the config rather than the stripe object! 
I think we can take this PR out of draft mode now :)

sirbrillig commented:
Cool. I want to do one more big rebase to clean it up, but that sounds good to me.

sirbrillig commented:
Moved some cleanup of `CreditCardPaymentBox` to #34694.

sirbrillig committed:
Add react-stripe-elements package

sirbrillig committed:
Add stripe as hidden payment method

sirbrillig committed:
Add StripeElementsPaymentBox

sirbrillig committed:
Support Stripe fields in CreditCardPaymentBox

sirbrillig committed:
CartStore: Add Stripe paymentDetails handling

sirbrillig committed:
TransactionStore: Save stripe object and config

sirbrillig committed:
Checkout actions: Add setStripeObject

sirbrillig committed:
Add Stripe helpers as lib/stripe

sirbrillig committed:
TransactionFlow: Support Stripe transactions

Add transaction handler for WPCOM_Billing_Stripe_Payment_Method

sirbrillig committed:
Add validation for Stripe elements form

sirbrillig committed:
wpcom-undocumented: Add stripeConfiguration

sirbrillig committed:
Checkout: Add pending redirect when order_id is set

In `Checkout.getCheckoutCompleteRedirectPath` if `order_id` is set, add `pending/order_id` to our redirect url.

sirbrillig commented:
I moved one other commit to its own PR, although maybe it wasn't necessary: #34720

Then I rebased all the remaining commits down as cleanly as I could. We should double check all the paths one more time before committing.

yoavf reviewed (approved):
👍  @sirbrillig ! I went through all the tests, and everything seems to be working great!
